java/src/main/java/com/google/appengine/tools/cloudstorage/dev/LocalRawGcsService.java

```diff
@@ -593,9 +630,27 @@ public ListItemBatch list(String bucket, String prefix, String delimiter,
     Set<String> prefixes = new HashSet<>();
     QueryResultIterator<Entity> dsResults =
         datastore.prepare(query).asQueryResultIterator(fetchOptions);
+    boolean fromBlobstore = false;
+    if (!dsResults.hasNext()) {
+      // If no results, perhaps there's metadata stored by the Blobstore API that we missed.
+      // Note that Blobstore metadata results aren't returned if some blobs are uploaded via
+      // Blobstore and others are uploaded using this client library.  We only check for Blobstore
+      // metadata if no results are found in appengine-gcs-client's metadata store.
+      Query blobstoreQuery = makeBlobstoreQuery();
+      dsResults = datastore.prepare(blobstoreQuery).asQueryResultIterator(fetchOptions);
+      fromBlobstore = true;
+    }
     while (items.size() < maxResults && dsResults.hasNext()) {
       Entity entity = dsResults.next();
-      String name = entity.getKey().getName();
+      String name;
+      if (!fromBlobstore) {
+        name = entity.getKey().getName();
+      } else {
+        String encodedName = entity.getKey().getName().substring(BLOBSTORE_KEY_PREFIX.length());
+        String fullName = new String(base64Url().omitPadding().decode(encodedName));
```

Comment:
As discussed, this may fail for blobs that were not uploaded to GCS.
Also, no bucket filtering?